File: app/windows/HyperWhisper/Services/Transcription/RustRetry.cs
// RUST RETRY WRAPPER (Wave 3 / Win-2)
//
// The retry policy is owned by the Rust core: `NextRetry(attempt, status, body,
// retryAfter)` classifies the (status, body) and returns a `RetryDecision` —
// `Retry(delayMs)` or `GiveUp`. This wrapper drives a single request through that
// decision loop via `RustHttpExecutor`, keeping ALL I/O, cancellation, and
// `Retry-After` header parsing on the platform side. Mirrors the macOS
// `RustRetry.swift`.
//
// Behavioral note (flagged in the PR): this unifies the previously-divergent
// per-provider retry loops onto the core's `NextRetry`. `NextRetry` is 1-based on
// the attempt that just FAILED and gives up at attempt >= RetryMaxAttempts()
// (== 8; exponential 1s, 2s, 4s, … 64s backoff), honoring `Retry-After` clamped
// to RetryMaxRetryAfterSecs() (== 10s). The core is RNG-free, so a small 0–30%
// jitter is added platform-side at the sleep point (see SleepAsync) to avoid a
// thundering herd. Poll loops in the multi-step providers do NOT go through this
// wrapper.
//
// 429/5xx vs transport split: the core's 8-attempt schedule is the GLOBAL bound
// and fully governs HTTP-status retries (429/5xx — the server is up, backing off
// is productive). Transport failures (no HTTP response at all) are additionally
// capped platform-side because grinding through all 8 attempts against a dead
// network stretches a doomed request to ~27 min:
//   - HttpRequestException (connection refused/reset, DNS): max 4 attempts —
//     restores the 1.7.0 ~14s failure envelope.
//   - Per-attempt timeout (OperationCanceledException with the CALLER token not
//     cancelled): max 2 attempts — the one useful retry is the one after
//     onTransportError rebuilds the pool; a second identical timeout means the
//     wait was already spent, fail fast.
// Under-cap transport failures still ask the core NextRetry(attempt, 503, ...)
// for the backoff delay, so the schedule stays core-owned.
//
// Backoff budget (issue #379): 8 attempts of raw exponential backoff is ~127s of
// sleep, so a hard-down provider used to take ~150s to fail. The core owns the
// rule (NextRetryWithinBudget); this driver just carries the running total of the
// delays the core handed it (sleptMs) back into the next decision. The core gives
// up when the next sleep would push that total past budgetMs (default
// RetryDefaultBudgetMs() == 30s), which stops at attempt 5 after 15s of backoff,
// before the 16/32/64s sleeps are reached. budgetMs: 0 means unbounded — exactly
// the old behaviour. This is a THIRD, orthogonal bound: MaxTransportAttempts and
// MaxTimeoutAttempts are per-failure-kind caps and are unchanged.
//
// sleptMs is BACKOFF ONLY — the time a failed request itself took is deliberately
// NOT charged to it. A large upload (GrokSttService allows a 5-min base / 30-min
// cap per attempt) would otherwise blow a 30s budget before its first 502 even
// arrived and get zero retries, and a 30s perAttemptTimeout would make
// MaxTimeoutAttempts below unreachable. Request duration is bounded by those
// per-attempt timeouts and by the two caps, not by the budget.
//
// TODO-verify (Windows/CI): Rust shared-core swap — compile-only; verify in CI.

using System.Net.Http;
using HyperWhisper.Models;
using uniffi.hyperwhisper_core;

namespace HyperWhisper.Services.Transcription;

internal static class RustRetry
{
    /// <summary>Max attempts that may end in a connection-level transport error
    /// (HttpRequestException) before failing terminally. See header comment.</summary>
    internal const int MaxTransportAttempts = 4;

    /// <summary>Max attempts that may end in a per-attempt timeout before failing
    /// terminally (i.e. one retry, giving the post-onTransportError fresh pool a
    /// single chance). See header comment.</summary>
    internal const int MaxTimeoutAttempts = 2;

    /// <summary>
    /// Source-compatible overload for providers with a fixed <see cref="HttpClient"/>
    /// for the life of the service. Services that can swap their client mid-sequence
    /// (DNS-recovery rebuild) must use the <c>Func&lt;HttpClient&gt;</c> shape so each
    /// attempt resolves the CURRENT client instead of pinning the pre-rebuild one.
    /// </summary>
    internal static Task<HttpResponse> PerformAsync(
        HttpClient client,
        Func<HttpRequest> buildRequest,
        Func<HttpResponse, TranscriptionException> parseError,
        CancellationToken cancellationToken,
        Func<Exception, Task>? onTransportError = null,
        TimeSpan? perAttemptTimeout = null,
        ulong? budgetMs = null)
    {
        return PerformAsync(() => client, buildRequest, parseError, cancellationToken, onTransportError, perAttemptTimeout, budgetMs);
    }

    /// <summary>
    /// Drive <paramref name="buildRequest"/>'s output through the executor + core
    /// retry loop.
    /// <list type="bullet">
    /// <item>On a 2xx response, returns the captured <see cref="HttpResponse"/>.</item>
    /// <item>On a non-2xx, parses <c>Retry-After</c> natively, asks the core
    /// <c>NextRetryWithinBudget(...)</c>, and either sleeps <c>delayMs</c> and
    /// retries or gives up.</item>
    /// <item>On a transport error with no HTTP response (network blip / timeout),
    /// treats it as a retryable 503-equivalent
    /// (<c>NextRetryWithinBudget(attempt, 503, "", null, …)</c>).</item>
    /// <item>On cancellation, throws <see cref="OperationCanceledException"/>.</item>
    /// <item>On give-up, throws the caller-mapped <see cref="TranscriptionException"/>
    /// derived from the last status/body (via <paramref name="parseError"/>), so
    /// callers surface the real failure rather than a generic one.</item>
    /// </list>
    /// <paramref name="buildRequest"/> is a delegate so the same
    /// <see cref="HttpRequest"/> is re-issued each attempt (the body is a file ref,
    /// so re-streaming is cheap and correct).
    ///
    /// <paramref name="onTransportError"/> is an OPTIONAL one-shot recovery hook
    /// invoked in the transport-error path BEFORE the next retry sleeps. Fired at
    /// most once per call (mirroring macOS' <c>didResetThisSequence</c> gate) so a
    /// flapping network can't thrash the pool.
    ///
    /// <paramref name="clientProvider"/> is resolved fresh on EVERY attempt so a
    /// service that swaps its <see cref="HttpClient"/> mid-sequence (e.g. the DNS
    /// rebuild in <c>onTransportError</c>) actually sends the next attempt on the
    /// new pool instead of the stale pre-rebuild one.
    ///
    /// <paramref name="perAttemptTimeout"/>, when set, bounds each individual
    /// attempt (each attempt gets the full budget); expiry counts against
    /// <see cref="MaxTimeoutAttempts"/>.
    ///
    /// <paramref name="budgetMs"/> bounds the TOTAL BACKOFF this sequence may
    /// sleep — not its wall clock, and explicitly not the time the failed
    /// requests themselves took. The core gives up rather than start a sleep that
    /// would push the running total past it. <c>null</c> = the core's interactive
    /// <c>RetryDefaultBudgetMs()</c> (30s); <c>0</c> = unbounded (the pre-#379
    /// behaviour), which a future batch caller can opt into.
    /// </summary>
    internal static async Task<HttpResponse> PerformAsync(
        Func<HttpClient> clientProvider,
        Func<HttpRequest> buildRequest,
        Func<HttpResponse, TranscriptionException> parseError,
        CancellationToken cancellationToken,
        Func<Exception, Task>? onTransportError = null,
        TimeSpan? perAttemptTimeout = null,
        ulong? budgetMs = null)
    {
        uint attempt = 0;
        // One-shot-per-sequence gate for the recovery hook.
        var didRecoverThisSequence = false;
        // Platform-side transport caps (see header comment). The core's
        // MAX_ATTEMPTS=8 remains the global bound across ALL failure kinds.
        var transportFailures = 0;
        var timeoutFailures = 0;
        var budget = budgetMs ?? HyperwhisperCoreMethods.RetryDefaultBudgetMs();
        // Running total of the backoff the core has asked this sequence to sleep.
        // Accumulated across the whole loop and never reset, so the budget bounds
        // the sequence rather than any single sleep.
        var sleptMs = 0UL;

        while (true)
        {
            cancellationToken.ThrowIfCancellationRequested();
            attempt += 1;

            var request = buildRequest();

            // Per-attempt timeout: linked CTS created INSIDE the loop so each
            // attempt gets the full budget (backoff sleeps run on the caller
            // token and must not consume the next attempt's budget).
            using var attemptCts = perAttemptTimeout.HasValue
                ? CancellationTokenSource.CreateLinkedTokenSource(cancellationToken)
                : null;
            attemptCts?.CancelAfter(perAttemptTimeout!.Value);
            var attemptToken = attemptCts?.Token ?? cancellationToken;

            HttpResponse response;
            try
            {
                response = await RustHttpExecutor
                    .ExecuteAsync(request, clientProvider(), attemptToken)
                    .ConfigureAwait(false);
            }
            // Ordering is load-bearing: genuine caller cancellation must be
            // tested against the CALLER token (the linked attempt token is also
            // cancelled when the per-attempt timeout fires) and never retried.
            catch (OperationCanceledException) when (cancellationToken.IsCancellationRequested)
            {
                throw;
            }
            catch (Exception ex) when (ex is HttpRequestException or OperationCanceledException)
            {
                // No HTTP response. Classify by exception type — the caller token
                // is NOT cancelled here (filtered above), so any cancellation
                // exception is the per-attempt timeout firing.
                var isTimeout = ex is OperationCanceledException;
                var failures = isTimeout ? ++timeoutFailures : ++transportFailures;
                var cap = isTimeout ? MaxTimeoutAttempts : MaxTransportAttempts;

                if (failures >= cap)
                {
                    LoggingService.Warn(
                        $"RustRetry: {(isTimeout ? "timeout" : "transport")} failure {failures}/{cap} on attempt {attempt} — giving up: {ex.Message}");
                    throw new TranscriptionException(
                        TranscriptionErrorCode.NetworkError,
                        isTimeout ? $"Request timed out after {failures} attempts: {ex.Message}" : ex.Message,
                        providerName: null,
                        innerException: ex);
                }

                // Under cap — the core still owns the backoff schedule; treat as a
                // retryable 503-equivalent.
                var decision = HyperwhisperCoreMethods.NextRetryWithinBudget(
                    @attempt: attempt,
                    @status: 503,
                    @body: "",
                    @retryAfter: null,
                    @sleptMs: sleptMs,
                    @budgetMs: budget);

                switch (decision)
                {
                    case RetryDecision.Retry retry:
                        LoggingService.Warn(
                            $"RustRetry: {(isTimeout ? "timeout" : "transport")} failure {failures}/{cap} on attempt {attempt} — retrying in {retry.@delayMs}ms: {ex.Message}");
                        if (!didRecoverThisSequence && onTransportError != null)
                        {
                            didRecoverThisSequence = true;
                            await onTransportError(ex).ConfigureAwait(false);
                        }
                        var admittedDelayMs = AdmittedSleepMs(retry.@delayMs, sleptMs, budget);
                        sleptMs += admittedDelayMs;
                        await SleepAsync(admittedDelayMs, cancellationToken).ConfigureAwait(false);
                        continue;

                    case RetryDecision.GiveUp:
                    default:
                        LogGiveUp(attempt, 503, "", null, sleptMs, budget);
                        throw new TranscriptionException(
                            TranscriptionErrorCode.NetworkError,
                            ex.Message,
                            providerName: null,
                            innerException: ex);
                }
            }

            // 2xx → success.
            if (response.@status is >= 200 and <= 299)
            {
                return response;
            }

            // Non-2xx → consult the core retry decision.
            var bodyText = System.Text.Encoding.UTF8.GetString(response.@body);
            var retryAfter = ParseRetryAfterHeader(response);

            // Floor at 0: a negative Retry-After (e.g. "-1") is meaningless and a
            // raw `(ulong)(-1)` would wrap to a huge delay → Task.Delay throws.
            var retryAfterMs = retryAfter.HasValue ? (ulong)Math.Max(0, retryAfter.Value) : (ulong?)null;

            var nonOkDecision = HyperwhisperCoreMethods.NextRetryWithinBudget(
                @attempt: attempt,
                @status: response.@status,
                @body: bodyText,
                @retryAfter: retryAfterMs,
                @sleptMs: sleptMs,
                @budgetMs: budget);

            switch (nonOkDecision)
            {
                case RetryDecision.Retry retry:
                    var admittedDelayMs = AdmittedSleepMs(retry.@delayMs, sleptMs, budget);
                    sleptMs += admittedDelayMs;
                    await SleepAsync(admittedDelayMs, cancellationToken).ConfigureAwait(false);
                    continue;

                case RetryDecision.GiveUp:
                default:
                    LogGiveUp(attempt, response.@status, bodyText, retryAfterMs, sleptMs, budget);
                    // The core's RateLimited carries no Retry-After (it doesn't
                    // read the header); enrich the give-up error with the value we
                    // parsed here so the "try again in N seconds" UI is preserved.
                    throw EnrichRateLimited(parseError(response), retryAfter);
            }
        }
    }

    /// <summary>
    /// Log a give-up with a reason a support engineer can act on: the backoff
    /// budget running out looks nothing like the attempt ceiling being hit, and
    /// the two want different fixes. Determined by re-asking the core WITHOUT the
    /// budget — if it would still have retried, the budget is what stopped us.
    /// One extra FFI call, only ever on the give-up path. Log line only: nothing
    /// downstream branches on this.
    /// </summary>
    private static void LogGiveUp(uint attempt, ushort status, string body, ulong? retryAfter, ulong sleptMs, ulong budgetMs)
    {
        var unbudgeted = HyperwhisperCoreMethods.NextRetry(
            @attempt: attempt,
            @status: status,
            @body: body,
            @retryAfter: retryAfter);

        LoggingService.Warn(unbudgeted is RetryDecision.Retry
            ? $"RustRetry: giving up on the retry BUDGET after attempt {attempt} — {sleptMs}ms of backoff slept of {budgetMs}ms (status {status})"
            : $"RustRetry: giving up on the retry POLICY after attempt {attempt} of {HyperwhisperCoreMethods.RetryMaxAttempts()} — {sleptMs}ms of backoff slept (status {status})");
    }

    /// <summary>
    /// When <paramref name="error"/> is a RateLimited with no RetryAfterSeconds,
    /// fill in the <c>Retry-After</c> value parsed from the response header.
    /// Otherwise pass the error through unchanged.
    /// </summary>
    private static TranscriptionException EnrichRateLimited(TranscriptionException error, int? retryAfter)
    {
        if (retryAfter.HasValue
            && error.Code == TranscriptionErrorCode.RateLimited
            && !error.RetryAfterSeconds.HasValue)
        {
            return new TranscriptionException(
                error.Code,
                error.Message,
                error.ProviderName,
                error.HttpStatusCode,
                // Clamp to ≥0 so a negative Retry-After can't surface a
                // "try again in -1 seconds" message to the user.
                Math.Max(0, retryAfter.Value),
                error.InnerException,
                error.ProviderDiagnostics);
        }
        return error;
    }

    /// <summary>
    /// Parse the integer <c>Retry-After</c> header from a binding
    /// <see cref="HttpResponse"/> (case-insensitive). Mirrors macOS
    /// <c>parseRetryAfterHeader</c>.
    /// </summary>
    private static int? ParseRetryAfterHeader(HttpResponse response)
    {
        foreach (var header in response.@headers)
        {
            if (string.Equals(header.@name, "Retry-After", StringComparison.OrdinalIgnoreCase))
            {
                if (int.TryParse(header.@value.Trim(), out var seconds))
                {
                    return seconds;
                }
                return null;
            }
        }
        return null;
    }

    /// <summary>
    /// Adds 0-30% jitter and caps the actual admitted sleep at the declared
    /// remaining budget. <paramref name="jitterUnit"/> is a pure test seam;
    /// production callers use <see cref="Random.Shared"/> across its full range.
    /// </summary>
    internal static ulong AdmittedSleepMs(
        ulong coreDelayMs,
        ulong sleptMs,
        ulong budgetMs,
        double? jitterUnit = null)
    {
        var unit = Math.Clamp(jitterUnit ?? Random.Shared.NextDouble(), 0, 1);
        var jittered = (ulong)(coreDelayMs * (1.0 + unit * 0.3));
        if (budgetMs == 0) return jittered;
        var remaining = budgetMs > sleptMs ? budgetMs - sleptMs : 0;
        return Math.Min(jittered, remaining);
    }

    private static Task SleepAsync(ulong admittedDelayMs, CancellationToken cancellationToken)
    {
        return Task.Delay(TimeSpan.FromMilliseconds(admittedDelayMs), cancellationToken);
    }
}

File: app/macos/hyperwhisper/Managers/Transcription/Support/RustRetry.swift
//
//  RustRetry.swift
//  hyperwhisper
//
//  Retry wrapper + boundary mappers for the Rust shared core (Wave 3 / M3-B).
//
//  The retry policy is owned by the core: `nextRetry(attempt:status:body:retryAfter:)`
//  classifies the (status, body) and returns a `RetryDecision` — `.retry(delayMs:)`
//  or `.giveUp`. This wrapper drives a single request through that decision loop
//  via `RustHTTPExecutor`, keeping ALL I/O, cancellation, and `Retry-After`
//  header parsing on the platform side.
//
//  Behavioral note (flagged in the PR): this unifies the previously-divergent
//  per-provider retry loops onto the core's `nextRetry`. `nextRetry` is 1-based
//  on the attempt that just FAILED and gives up at attempt >= MAX_ATTEMPTS (8),
//  with exponential backoff (1s, 2s, 4s, … 64s), honoring `Retry-After` clamped
//  to 10s. The core is RNG-free, so a small randomized jitter (0–30%) is added
//  platform-side at the sleep point (see `sleep(_:)`) to avoid a thundering herd.
//

import Foundation

enum RustRetry {

    /// Drive `buildRequest()`'s output through the executor + core retry loop.
    ///
    /// - On a 2xx response, returns the captured `HttpResponse`.
    /// - On a non-2xx, parses `Retry-After` natively, asks the core
    ///   `nextRetry(...)`, and either sleeps `delayMs` and retries or gives up.
    /// - On a `URLError` with no HTTP response (network blip), treats it as a
    ///   retryable 503-equivalent (`nextRetry(attempt, 503, "", nil)`).
    /// - On cancellation, throws `CancellationError`.
    /// - On give-up, throws the core-mapped Swift `TranscriptionError` derived
    ///   from the last status/body (via `parseError`), so callers surface the
    ///   real failure rather than a generic one.
    ///
    /// `buildRequest` is a closure so the same `HttpRequest` is re-issued each
    /// attempt (the body is a file ref, so re-streaming is cheap and correct).
    ///
    /// `onTransportError` is an OPTIONAL one-shot recovery hook invoked in the
    /// transport-error path (a `URLError` with no HTTP response) BEFORE the next
    /// retry sleeps. It is fired at most once per `perform(...)` call (mirroring
    /// the original `performRequestWithRetry`'s `didResetThisSequence` gate) so a
    /// flapping network can't thrash the pool. HyperWhisper Cloud / routed pass a
    /// DNS-shaped session-reset closure here so a network flip
    /// (VPN/captive-portal/tether swap) re-resolves instead of burning all
    /// attempts against a poisoned connection pool. Default `nil` = no-op, so
    /// other callers are unaffected.
    static func perform(
        session: URLSession,
        buildRequest: () throws -> HttpRequest,
        parseError: (HttpResponse) -> TranscriptionError,
        onTransportError: ((URLError) async -> Void)? = nil
    ) async throws -> HttpResponse {
        var attempt: UInt32 = 0
        // One-shot-per-sequence gate for the recovery hook (matches the original
        // native `didResetThisSequence` semantics).
        var didRecoverThisSequence = false

        while true {
            try Task.checkCancellation()
            attempt += 1

            let request = try buildRequest()

            // Perform one attempt. A thrown error here is either cancellation or
            // a URLSession transport error (no HTTP response).
            let response: HttpResponse
            do {
                response = try await RustHTTPExecutor.execute(request, session: session)
            } catch is CancellationError {
                throw CancellationError()
            } catch {
                let nsError = error as NSError
                if nsError.domain == NSURLErrorDomain && nsError.code == NSURLErrorCancelled {
                    throw CancellationError()
                }

                // No HTTP response — treat as a retryable 503-equivalent.
                let decision = nextRetry(attempt: attempt, status: 503, body: "", retryAfter: nil)
                switch decision {
                case let .retry(delayMs):
                    // One-shot transport-error recovery (e.g. DNS-cache flush on a
                    // network flip) BEFORE the retry sleeps. Gated to once per
                    // sequence so a flapping network can't thrash the pool.
                    if !didRecoverThisSequence, let hook = onTransportError, let urlError = error as? URLError {
                        didRecoverThisSequence = true
                        await hook(urlError)
                    }
                    try await sleep(delayMs)
                    continue
                case .giveUp:
                    throw TranscriptionError.transientNetwork(details: error.localizedDescription)
                }
            }

            // 2xx → success.
            if (200...299).contains(Int(response.status)) {
                return response
            }

            // Non-2xx → consult the core retry decision.
            let bodyText = String(data: response.body, encoding: .utf8) ?? ""
            let retryAfter = parseRetryAfterHeader(response)

            let decision = nextRetry(
                attempt: attempt,
                status: response.status,
                body: bodyText,
                // Clamp at the conversion: `parseRetryAfterHeader` uses `Int(...)`
                // and so accepts negatives; `UInt64(-1)` would TRAP. A negative
                // Retry-After is meaningless, so floor it at 0. (The `Int?` value
                // is still passed to `enrichRateLimited` below for user messaging.)
                retryAfter: retryAfter.map { UInt64(max(0, $0)) }
            )

            switch decision {
            case let .retry(delayMs):
                try await sleep(delayMs)
                continue
            case .giveUp:
                // The core's RateLimited carries no Retry-After (it doesn't read
                // the header); enrich the give-up error with the value we parsed
                // here so the "try again in N seconds" UI is preserved.
                throw enrichRateLimited(parseError(response), retryAfter: retryAfter)
            }
        }
    }

    /// When `error` is a `.rateLimited` with a nil `retryAfter`, fill in the
    /// `Retry-After` value parsed from the response header. Otherwise pass the
    /// error through unchanged.
    private static func enrichRateLimited(_ error: TranscriptionError, retryAfter: Int?) -> TranscriptionError {
        guard let retryAfter, case .rateLimited(let existing) = error, existing == nil else {
            return error
        }
        return .rateLimited(retryAfter: retryAfter)
    }

    /// Parse the integer `Retry-After` header from a binding `HttpResponse`,
    /// reading the header list the core captured (case-insensitive).
    private static func parseRetryAfterHeader(_ response: HttpResponse) -> Int? {
        guard let value = response.headers.first(where: {
            $0.name.caseInsensitiveCompare("Retry-After") == .orderedSame
        })?.value else { return nil }
        return Int(value.trimmingCharacters(in: .whitespaces))
    }

    private static func sleep(_ delayMs: UInt64) async throws {
        // Add randomized jitter (0–30%) on top of the core's deterministic
        // backoff so concurrent clients don't all retry in lockstep (thundering
        // herd). The core forbids RNG, so the jitter lives here — restoring the
        // pre-migration `.transcription` preset's `0...0.3` jitterRange.
        let jitterFactor = 1.0 + Double.random(in: 0...0.3)
        let nanos = UInt64(Double(delayMs) * jitterFactor * 1_000_000)
        try await Task.sleep(nanoseconds: nanos)
    }
}

// MARK: - Boundary mappers

enum RustCoreMapping {

    /// The standard `parseError` give-up closure for `RustRetry.perform`.
    ///
    /// On a non-2xx response, runs the provider's core parser (which classifies
    /// the body into an `HwTranscriptionError`) and maps the result to a Swift
    /// `TranscriptionError` tagged with `providerName`. A non-throwing parse
    /// (unexpected on a non-2xx) and any non-core error both fall back to
    /// `.invalidResponse`. This is the dedup of the identical inline closures
    /// every provider used to carry; providers that need extra context on the
    /// give-up error (HW Cloud / routed 402-credit + file-size enrichment) keep
    /// their own bespoke closure.
    static func parseErrorClosure(
        providerName: String,
        _ parse: @escaping (HttpResponse) throws -> Void
    ) -> (HttpResponse) -> TranscriptionError {
        { resp in
            do {
                try parse(resp)
                return TranscriptionError.invalidResponse(details: "unexpected non-error response")
            } catch let err as HwTranscriptionError {
                return RustCoreMapping.mapTranscriptionError(err, providerName: providerName)
            } catch {
                return TranscriptionError.invalidResponse(details: error.localizedDescription)
            }
        }
    }

    /// Map a core `HwTranscriptionError` to the app's `TranscriptionError`.
    ///
    /// `providerName` is the display name for messaging; `creditsRemaining` /
    /// `creditsRequired` carry the credit context for `.insufficientCredits`
    /// (HW Cloud quota path), when the platform read them from the response body.
    /// `fileTooLargeBytes` / `fileTooLargeLimit` carry the real audio/limit sizes
    /// for `.audioFileTooLarge` (HW Cloud / routed 413), when the platform parsed
    /// `actual_size_mb` / `max_size_mb` from the response body. Default 0 = absent.
    static func mapTranscriptionError(
        _ error: HwTranscriptionError,
        providerName: String,
        insufficientCredits: Bool = false,
        creditsRemaining: Int = 0,
        creditsRequired: Int = 0,
        fileTooLargeBytes: Int64 = 0,
        fileTooLargeLimit: Int64 = 0
    ) -> TranscriptionError {
        switch error {
        case .Unauthorized:
            return .unauthorized(provider: providerName)
        case .QuotaExceeded:
            // HW Cloud / routed: a 402 is "out of credits". Prefer the richer
            // .insufficientCredits when the caller pulled the credit context.
            if insufficientCredits {
                return .insufficientCredits(remaining: creditsRemaining, required: creditsRequired)
            }
            return .quotaExceeded(provider: providerName, message: nil)
        case .FileTooLarge:
            return .audioFileTooLarge(fileSize: fileTooLargeBytes, limit: fileTooLargeLimit, providerName: providerName)
        case let .RateLimited(retryAfterSecs):
            return .rateLimited(retryAfter: retryAfterSecs.map { Int($0) })
        case let .ProviderUnavailable(status):
            return .serverError(statusCode: Int(status), message: "Provider unavailable")
        case .NoSpeech:
            return .noSpeechDetected
        case let .BadRequest(status, message):
            // 400-class. Surface as a server error carrying the upstream message.
            // Even when the message is empty, preserve the HTTP status (matching
            // Windows `RustCoreMapping.cs`) rather than collapsing to a statusless
            // .invalidRequest — the status is diagnostic.
            if message.isEmpty {
                return .serverError(statusCode: Int(status), message: "Invalid request")
            }
            return .serverError(statusCode: Int(status), message: message)
        case let .Parse(message):
            return .invalidResponse(details: message)
        }
    }

    /// Map the app's `CloudProvider`-style routed header to an `HwProvider`.
    /// Used by routed + (later) health probes.
    static func hwProvider(for sttProviderHeader: String) -> HwProvider {
        switch sttProviderHeader {
        case "azure-mai": return .azureMai
        case "google-chirp": return .googleChirp
        default: return .hyperWhisperCloud
        }
    }

    /// Map the app's full `CloudProvider` enum to an `HwProvider`. Used by the
    /// health probes (`buildHealthRequest`/`parseHealthResponse`).
    ///
    /// The three HW-Cloud-routed providers map onto the core's routed cases
    /// (`hyperwhisper`→`hyperWhisperCloud`, `microsoftAzureSpeech`→`azureMai`,
    /// `googleSpeech`→`googleChirp`); every direct vendor maps 1:1.
    static func hwProvider(for provider: CloudProvider) -> HwProvider {
        switch provider {
        case .hyperwhisper: return .hyperWhisperCloud
        case .openai: return .openai
        case .groq: return .groq
        case .deepgram: return .deepgram
        case .assemblyAI: return .assemblyai
        case .elevenLabs: return .elevenlabs
        case .mistral: return .mistral
        case .soniox: return .soniox
        case .gemini: return .gemini
        case .grok: return .grok
        case .microsoftAzureSpeech: return .azureMai
        case .googleSpeech: return .googleChirp
        }
    }

    /// Build a core `TranscribeParams` from the platform's transcription inputs.
    ///
    /// The core builds the vocabulary CSV itself from `vocabulary` (trim +
    /// drop-empty, no lowercase/dedup) — pass the RAW term list, do NOT
    /// pre-encode. `audioMime` is passed explicitly from
    /// `AudioMimeTypeResolver` rather than letting the core re-resolve.
    static func transcribeParams(
        audioPath: String,
        audioMime: String,
        language: String?,
        vocabulary: [String],
        apiKey: String = "",
        model: String = "",
        prompt: String? = nil,
        baseURL: String? = nil,
        licenseKey: String? = nil,
        deviceID: String? = nil,
        routedProvider: String? = nil,
        routedModel: String? = nil,
        routedDomain: String? = nil
    ) -> TranscribeParams {
        TranscribeParams(
            apiKey: apiKey,
            model: model,
            language: language,
            vocabulary: vocabulary,
            prompt: prompt,
            temperature: nil,
            audioPath: audioPath,
            audioMime: audioMime,
            baseUrl: baseURL,
            licenseKey: licenseKey,
            deviceId: deviceID,
            routedProvider: routedProvider,
            routedModel: routedModel,
            routedDomain: routedDomain
        )
    }

    /// Extract vocabulary boost terms for provider egress.
    ///
    /// Replacement-pair entries are post-transcription substitutions, not
    /// recognition hints. The Rust core applies the same sanitizer/dedup/caps
    /// again while building provider requests, but pre-sanitizing here keeps the
    /// FFI boundary free of replacement-pair and oversized prompt-injection data.
    static func boostVocabularyTerms(from vocabulary: [Vocabulary]) -> [String] {
        var terms: [String] = []
        var seen = Set<String>()
        for item in vocabulary {
            guard item.replacement?.isEmpty != false else { continue }
            guard let raw = item.word else { continue }
            let sanitized = PromptBuilder.sanitizeVocabularyWord(raw)
            guard !sanitized.isEmpty else { continue }
            guard seen.insert(sanitized.lowercased()).inserted else { continue }
            terms.append(sanitized)
        }
        return terms
    }

    /// Parse the HW-Cloud / routed 413 size context (`actual_size_mb` /
    /// `max_size_mb`) from an error response body into bytes, matching the old
    /// native routed `handleHTTPError`. Returns `(0, 0)` when absent.
    static func fileTooLargeContext(from response: HttpResponse) -> (bytes: Int64, limit: Int64) {
        guard let json = try? JSONSerialization.jsonObject(with: response.body) as? [String: Any] else {
            return (0, 0)
        }
        let limit = (json["max_size_mb"] as? Int).map { Int64($0) * 1_048_576 } ?? 0
        let bytes = (json["actual_size_mb"] as? Double).map { Int64($0 * 1_048_576) } ?? 0
        return (bytes, limit)
    }
}
